### DISCUSSION TITLE
Neat work and comments from Glaze designer

### DISCUSSION THREAD

**Discussion by stephenberry** [RealTimeChris/Jsonifier#26] [General]
Hey, I came across your library and it's neat what you've done with Glaze and simdjson. That's quite a bit of work to rewrite it all.

I'm always interested in improving the performance in Glaze, and it looks like you might have found some improvements using SIMD. However, it's hard for me to tell exactly where the performance improvements are coming from, especially because some things in Jsonifier for writing or parsing are unsafe. And, on platforms where SIMD isn't supported, like Arm Macs, Glaze is significantly faster.

Note that your parsing of booleans is not JSON compliant because you check the first character for `t` or `f`, but if that passes you only compare against `true`, so if the input is `fluge`, it will read as false even though this is invalid JSON.

I'm not sure if you're benchmarking with MSVC, but it's worth noting that MSVC is not nearly as good a compiler as Clang and GCC right now and Glaze is significantly faster (sometimes twice as fast) with GCC than MSVC on the same platform.
Could you share what compiler you're using for your benchmarks?

I'm curious what you mean exactly by resaturating the CPU registers for better SIMD performance? Care to elaborate?

Neat work! I hope to learn from you.

Sincerely,
Stephen

**Comment by RealTimeChris** [comment]
My bad I never saw this when you posted it. I will update the response once I've gotten to it. 

What I mean by resaturating the CPU registers instead of only operating on 64-bytes worth of string data at a time refers to the phase-1 parsing done within Simdjson, where they use SIMD instructions to first collect the structural indices of the json data. Instead of repacking the data into SIMD registers after the first part of phase-1, they begin to operate using 64-bit registers on the string data. As far as I could tell, repacking the data into SIMD registers seems to offer a roughly 30% increase in performance for phase-1 of parsing.
And yes I was using MSVC for performance benchmarking at the time you wrote this discussion but I have since switched to using CLANG-18. 

Additionally, I've since improved on the unsafe bool parsing methods you mentioned here.

Cheers!

**Reply by stephenberry** [reply]
No problem, I've noticed that you've addressed most of my initial concerns. I will mention that there are some corner case numeric parsing issues that were fixed in Glaze since you adopted the code (such as the minimum int64_t and floating point zeros with massive numbers of digits). You might check out how Glaze handles these cases.

Repacking the data into SIMD registers is a neat approach. Thanks for the explanation.

And, thanks for all your research into performant JSON!
